core/src/main/java/org/apache/lucene/search/vectorhighlight/CustomFieldQuery.java

```diff
@@ -56,6 +58,11 @@ public CustomFieldQuery(Query query, IndexReader reader, boolean phraseHighlight
 
     @Override
     void flatten(Query sourceQuery, IndexReader reader, Collection<Query> flatQueries, float boost) throws IOException {
+        while (sourceQuery instanceof BoostQuery) {
```

Comment:
I'm not sure why you used `while` here instead of `if` like the rest of the cases. Do we usually have multiple layers of `BoostQuery` when we have a single layer? Could you push a patch with a comment about it or switch it to the normal `if` statement?